### PR TITLE
PK checks

### PR DESCRIPTION
**Pull request by bugthunk**
Add options and functionality to enforce PKs by means of checks.

An app branch to test against will be provided in private chat.

### CODE REVIEW

**Comment by bugthunk** [comment]
While we're at it we should use a DT to collect the different options so we don't get a proliferation of config parameters. I'll do it tomorrow. @jonathanjouty @23Skidoo

File: src/Database/PostgreSQL/PQTypes/Checks.hs [hunk 1]
```diff
@@ -381,6 +383,9 @@ checkDBStructure tables = fmap mconcat . forM tables $ \table ->
         checkPrimaryKey mdef mpk = mconcat [
             checkEquality "PRIMARY KEY" def (map fst pk)
           , checkNames (const (pkName tblName)) pk
+          , if (EnforceMandatoryPrimaryKeys `elem` options)
```
Comment (23Skidoo) [inline src/Database/PostgreSQL/PQTypes/Checks.hs]: I'm wondering whether we should ditch this pattern. More standard way is to do something like

```
data CheckOptions = CheckOptions {
    chkOptEnforceMandatoryPrimaryKeys :: Bool,
    ...
}
```

File: src/Database/PostgreSQL/PQTypes/Checks/Util.hs [hunk 2]
```diff
@@ -85,6 +86,26 @@ checkNames prop_name = mconcat . map check
           , ")."
           ]]
 
+checkPresence :: RawSQL ()
```
Comment (23Skidoo) [inline src/Database/PostgreSQL/PQTypes/Checks/Util.hs]: `checkPKPresence`? Also needs Haddock comments for both the top-level entity and the parameters.

**Comment by 23Skidoo** [comment]
Also looks like the test suite is broken by these changes.

File: src/Database/PostgreSQL/PQTypes/Checks.hs (hunk 1 above)
Comment (23Skidoo) [reply src/Database/PostgreSQL/PQTypes/Checks.hs]: I see, you already wrote above that you want to do this.

File: src/Database/PostgreSQL/PQTypes/Checks/Util.hs (hunk 2 above)
Comment (bugthunk) [reply src/Database/PostgreSQL/PQTypes/Checks/Util.hs]: Addressed. Fixing tests now.

**Comment by bugthunk** [comment]
@23Skidoo Tests should be good now. I'm not entirely satisfied with the `ExtrasOptions` name, but I didn't want to waste time thinking about this. I also refactored some `{- -}`-style explanations of what the arguments are in `test/Main.hs` into using `let`-bindings with explanatory variable names instead.

**Comment by bugthunk** [comment]
...aaand @jonathanjouty and @23Skidoo: While I've tested this against our system, doing all combinations would take too long. I've done the important ones (table def'd without PK, and with) and it works as intended - there aren't many contortions in there, so it should just work. I guess what I'm trying to say is that I'm not happy about test coverage - neither manual nor automated.

We could wait and discuss this before a potential merge. Perhaps it's good if at least one other person does some manual testing (it's a hassle but it would at least make me feel better 😸 ).

**Comment by 23Skidoo** [comment]
Can we add an automated test case then?

**Comment by bugthunk** [comment]
Yes, we should... I'd have to do it when I don't feel inundated by starting up our new project. OTOH merging this isn't pressing.

**Comment by 23Skidoo** [comment]
@bugthunk If you want, we can merge this without tests, but let's at least make a ticket for adding them at some point in the future.

**Comment by bugthunk** [comment]
@23Skidoo I'd like @jonathanjouty to check and 'aye' it as well in that case. Other than that I agree with you wholeheartedly.

**Comment by bugthunk** [comment]
Working on some tests.

**Comment by bugthunk** [comment]
@23Skidoo @jonathanjouty I added some tests.

When and if you run them, you'll see something along the lines of
```
2018-03-02 12:19:17 ATTENTION hpqtypes-extras-test: There are migration chains selected for execution that expect a different starting table version number from the one in the database. This likely means that the order of migrations is wrong. {
  "tables": [
    "bad_guy",
    "bank",
    "participated_in_robbery",
    "robbery",
    "under_arrest"
  ]
}
```
I don't know what that is, and furthermore it's been there since at least `v1.4.0.0`.

**Comment by 23Skidoo** [comment]
I'll look into it, since it was me who added that test.

**Comment by bugthunk** [comment]
Checking Travis errors.

**Comment by bugthunk** [comment]
@23Skidoo @jonathanjouty Errors went away when I ran the job again.

**Pull request merged**